src/Microsoft.AspNetCore.Server.Kestrel/KestrelWebHostBuilderExtensions.cs

```diff
@@ -1,16 +1,31 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
 using System.Reflection;
+using Microsoft.AspNetCore.Hosting.Server;
 using Microsoft.AspNetCore.Server.Kestrel;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
 
 namespace Microsoft.AspNetCore.Hosting
 {
     public static class KestrelWebHostBuilderExtensions
     {
         public static IWebHostBuilder UseKestrel(this IWebHostBuilder hostBuilder)
         {
-            return hostBuilder.UseServer(typeof(KestrelServer).GetTypeInfo().Assembly.FullName);
+            return hostBuilder.ConfigureServices(services => services.AddSingleton<IServerFactory, ServerFactory>());
+        }
+
+        public static IWebHostBuilder UseKestrel(this IWebHostBuilder hostBuilder, Action<KestrelServerOptions> options)
```

Comment:
We want to split server configuration from application configuration. Of course the user can also configure options in their startup class but this is the default experience we want to preserve. It also keeps the server specific stuff in Program.Main instead of putting it in the user's Startup class by default.